teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,420 @@
+
+#### `uri`
+
+The `uri` component identifies an artifact within the artifact storage in use.
+The storage itself is included as part of the `TaskRun` and `PipelineRun` definition.
+
+The `uri` is in [purl-generic](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#generic). Example values:
+
+- `pkg:generic/<name>`
+- `pkg:generic/build`
+- `pkg:generic/repo`
+- `pkg:generic/test-results`
+- `pkg:generic/sbom`
+
+
+#### `digest`
+
+The `digest` component includes the result of applying an hash function to the artifact itself.
+The `digest` is in the format `hash-function:<digest>`. Example values:
+
+- `md5:471a3a9ddc1d6e699190b9f4ca88aadf`
+- `sha256:8796357729cfd877cf8fa7d45a8ab3524d9249c23a0bf68bb0026c0783b881d2`
+
+### Steps Implementation
+
+Steps in Tekton do not have a YAML-based API. This may change one [TEP-0142 - Reusable Steps](https://github.com/tektoncd/community/pull/1065) is implemented, but only for reusable steps. Inlined steps will continue same as today.
+
+Even without a YAML API, the step interface can be defined as a combination of input environment variables and files and output files.
+Tekton will then map data coming from the Tekton API into the `download` step interface and data produced by the `upload` and `download` steps into the `TaskRun` status.
+
+This approach enables users who do not wish to use the [artifacts API](#artifacts-api) to use custom steps and still have provenance data injected into the `TaskRun` status.
+
+Steps are implemented as a single go binary that takes command line arguments to implement the logic of the four different steps.
+The go binary is built for all Tekton supported architectures and OSes. The steps point to the multi-arch image manifest so that the container runtime may download the correct version.
+
+#### Digest
+
+```
+[`ARTIFACT_NAMES`] -> [`/tekton/artifacts/uploads/<digest>.tgz`, `/tekton/artifacts/uploads/<digest>.json`]
+```
+
+The `digest` step creates and archive and calculates the digest of an array of artifacts from the local `${ARTIFACT_OUTPUTS}:-/tekton/artifacts/outputs/}` storage, based on a list of `names`:
+- when the `ARTIFACT_NAMES` environment variables is set, the `digest` step searches for each `name` in the `$ARTIFACT_OUTPUTS` folder, and fail if at least one is not found
+
+```mermaid
+flowchart LR
+  start(( start )) --> loop
+  loop[[for each name in\n $ARTIFACT_NAMES]] --> check{folder exists\n$ARTIFACT_OUTPUTS/\n$ARTIFACT_NAMES}
+  check -- no --> failure(( failure ))
+  check -- yes --> tarball[create a .tar.gz\n+\ntarball digest]
+  tarball --> meta[append to digest.json:\n`uri`, `digest`]
+  meta --> loop
+  meta --> done(( success ))
+```
+
+- when the `ARTIFACT_NAMES` environment variables is not set, the `digest` step considers any file or folder in the `$ARTIFACT_OUTPUTS` folder as an artifact. This enables use cases where the list of artifact is not known upfront.
```

Comment:
I think we would run into the following issue here:

If users simply make the data available at  `$ARTIFACT_OUTPUTS` then how do they inform the Tekton Controller about their desired storage location? What if they are not ok with the PVC based storage and instead want to upload to GCS? How do they inform their need to the Tekton controller without the Artifact API in the first place?

Instead, if we have it such that without the Artifact API, they have to write their steps to do all the work (i.e. archiving, uploading of Artifacts and making the metadata available at `/tekton/artifacts/uploads/<digest>.json`). The expectation from the users is to populate `/tekton/artifacts/uploads/<digest>.json` instead of making the artifacts available at `$ARTIFACT_OUTPUTS`. The digest binary would then just look for the metadata of the artifact and skip the archiving and upload part. 

With the Artifact API, they would be able to (probably in the future) say which reusable Step they want to back the artifact with. Without that, they need to do all the work themselves because the Controller cannot get any input about the user's desire.

Reply:
> If users simply make the data available at $ARTIFACT_OUTPUTS then how do they inform the Tekton Controller about their desired storage location? What if they are not ok with the PVC based storage and instead want to upload to GCS? How do they inform their need to the Tekton controller without the Artifact API in the first place?

Alternatives to the storage location are out of scope for this TEP, but I understand we should design with them in mind. 
My idea is that the backend storage is associated with the specific step implementation provided by the user for the `upload` and `download` steps. 

Custom user steps might might need some extra inputs to be used (name of the bucket, path to credentials, etc). Some of these may be platform wide, others may be `PipelineRun` specific. I would say that good step implementation should ideally not need `PipelineRun` specific inputs - it might use the `PipelineRun` name or uuid for instance where needed.

Reply:
> Without the Artifact API, they should to do all the work themselves because the Controller cannot get any input about the user's desire.

More work yet, but not *all* the work:
- the step interface is designed so that if a step implements it, the Tekton controller can still get the provenance from local disk, through termination message/sidecar log, into the status
- downstream tasks could still consume that data from previous tasks, but since they would have no knowledge upfront, they might need a more complex logic in their script to parse and filter upstream artifacts and decide which to use

I tried to explain this in the paragraphs above, but I can present this again tomorrow during the meeting and answers more questions.